File: jsol-compiler-php.bak/0100-cli-parser.php
<?php
// @JSOL v0.2.97 - CLI Arguments Parser (generic, target-agnostic)
//
// No target id is hardcoded here. Any "--<id>-target=", "--<id>-prefix=",
// "--<id>-suffix=" flag is parsed generically into $mOptions[id + "Target"/
// "Prefix"/"Suffix"], for ANY id — adding a new compiler target never
// requires touching this file again.

// Contains the ONLY place "py" gets normalized to "python" in the whole compiler.
// "python" is canonical everywhere else (matches targets/python/rules.json,
// $mSSOT["targets"]["python"], $mBackendRegistry). "py" survives only as the
// CLI's short spelling and the output file extension (handled elsewhere).


$bCli_EndsWith = function ($saStr, $saSuffix) {
  $iStrLen = mb_strlen($saStr, "UTF-8");
	$iSufLen = mb_strlen($saSuffix, "UTF-8");
	if ($iSufLen > $iStrLen) {
    return false;
  }
  return mb_substr($saStr,  $iStrLen - $iSufLen,  $iSufLen, "UTF-8") === $saSuffix;
};
$saCli_NormalizeTargetId = function ($saId) {
  if ($saId === "py") {
    return "python";
  }
  return $saId;
};
$mParseRawCliArgs = function($aRawArgs) use (&$bCli_EndsWith, &$saCli_NormalizeTargetId) {
  $mOptions = JSOL::dict("source",  "",  "sourceDir",  "",  "outDir",  "",  "target",  "",  "targets",  "");
	$iCount = count($aRawArgs);

	for ($i = 0; $i < $iCount; $i = $i + 1) {
    $saArg = $aRawArgs[$i];
		if (Str::indexOf($saArg,  "--") === 0) {
      $saClean = mb_substr($saArg,  2,  mb_strlen($saArg, "UTF-8") - 2, "UTF-8");
			$iEqIndex = Str::indexOf($saClean,  "=");
			$saKey = $saClean;
			$saVal = "true";

			if ($iEqIndex !== -1) {
        $saKey = mb_substr($saClean,  0,  $iEqIndex, "UTF-8");
				$saVal = mb_substr($saClean,  $iEqIndex + 1,  mb_strlen($saClean, "UTF-8") - ($iEqIndex + 1), "UTF-8");
      }
      if ($saKey === "source") {
        $mOptions["source"] = $saVal;
      }
      else if ($saKey === "source-dir") {
        $mOptions["sourceDir"] = $saVal;
      }
      else if ($saKey === "out-dir") {
        $mOptions["outDir"] = $saVal;
      }
      else if ($saKey === "targets") {
        $mOptions["targets"] = $saVal;
      }
      else if ($saKey === "target") {
        $mOptions["target"] = $saVal;
      }
      else if ($bCli_EndsWith($saKey, "-target") === true) {
        $mOptions[Str::concat($saCli_NormalizeTargetId(mb_substr($saKey,  0,  mb_strlen($saKey, "UTF-8") - 7, "UTF-8")),  "Target")] = $saVal;
      }
      else if ($bCli_EndsWith($saKey, "-prefix") === true) {
        $mOptions[Str::concat($saCli_NormalizeTargetId(mb_substr($saKey,  0,  mb_strlen($saKey, "UTF-8") - 7, "UTF-8")),  "Prefix")] = $saVal;
      }
      else if ($bCli_EndsWith($saKey, "-suffix") === true) {
        $mOptions[Str::concat($saCli_NormalizeTargetId(mb_substr($saKey,  0,  mb_strlen($saKey, "UTF-8") - 7, "UTF-8")),  "Suffix")] = $saVal;
      }
    }
  }
  return $mOptions;
};

File: jsol-compiler-py.bak/0100-cli-parser.py
import math
import functools
from jsol_core import JSOL

# @JSOL v0.2.97 - CLI Arguments Parser (generic, target-agnostic)
#
# No target id is hardcoded here. Any "--<id>-target=", "--<id>-prefix=",
# "--<id>-suffix=" flag is parsed generically into $mOptions[id + "Target"/
# "Prefix"/"Suffix"], for ANY id — adding a new compiler target never
# requires touching this file again.

# Contains the ONLY place "py" gets normalized to "python" in the whole compiler.
# "python" is canonical everywhere else (matches targets/python/rules.json,
# $mSSOT["targets"]["python"], $mBackendRegistry). "py" survives only as the
# CLI's short spelling and the output file extension (handled elsewhere).


def bCli_EndsWith(saStr, saSuffix): 

  iStrLen = len(saStr);
  iSufLen = len(saSuffix);
  if iSufLen > iStrLen: 

    return False;


  return saStr[( iStrLen - iSufLen):( iStrLen - iSufLen)+( iSufLen)] == saSuffix;


def saCli_NormalizeTargetId(saId): 

  if saId == "py": 

    return "python";


  return saId;


def mParseRawCliArgs(aRawArgs): 

  mOptions = JSOL.dict("source",  "",  "sourceDir",  "",  "outDir",  "",  "target",  "",  "targets",  "");
  iCount = len(aRawArgs);

  i = 0;
  while i < iCount: 

    saArg = aRawArgs[i];
    if JSOL.str_index_of(saArg,  "--") == 0: 

      saClean = saArg[( 2):( 2)+( len(saArg) - 2)];
      iEqIndex = JSOL.str_index_of(saClean,  "=");
      saKey = saClean;
      saVal = "true";

      if iEqIndex != -1: 

        saKey = saClean[( 0):( 0)+( iEqIndex)];
        saVal = saClean[( iEqIndex + 1):( iEqIndex + 1)+( len(saClean) - (iEqIndex + 1))];


      if saKey == "source": 

        mOptions["source"] = saVal;


      elif saKey == "source-dir": 

        mOptions["sourceDir"] = saVal;


      elif saKey == "out-dir": 

        mOptions["outDir"] = saVal;


      elif saKey == "targets": 

        mOptions["targets"] = saVal;


      elif saKey == "target": 

        mOptions["target"] = saVal;


      elif bCli_EndsWith(saKey, "-target") == True: 

        mOptions["".join(JSOL.to_str(_x) for _x in [saCli_NormalizeTargetId(saKey[( 0):( 0)+( len(saKey) - 7)]),  "Target"])] = saVal;


      elif bCli_EndsWith(saKey, "-prefix") == True: 

        mOptions["".join(JSOL.to_str(_x) for _x in [saCli_NormalizeTargetId(saKey[( 0):( 0)+( len(saKey) - 7)]),  "Prefix"])] = saVal;


      elif bCli_EndsWith(saKey, "-suffix") == True: 

        mOptions["".join(JSOL.to_str(_x) for _x in [saCli_NormalizeTargetId(saKey[( 0):( 0)+( len(saKey) - 7)]),  "Suffix"])] = saVal;




    i = i + 1;


  return mOptions;


